### PR TITLE
ENH: allow @property decorator on external ctypedef classes

### PR DESCRIPTION
**Pull request by mattip**
Following up the discussion on the mailing list. I got to the point where I can use the new syntax:

- [X] Write a srctree test that creates various types of external ctypedef classes, add one that overrides a python-level @property with a cython one. This is a continuation of the test added in #2629 
- [X] Add `decorator` child and outer attributes to `CFuncDefNode`. ~I simply copied this from `DefNode`, is it required?~ - yes, seems to be required
- [X] ~Add a `DecoratorTransform.visit_CFuncDefNode` function based on `DecoratorTransform.visit_DefNode` and verify it is reached while parsing the `ctypedef class`.~
- [x] ~Found the point at which the attribute lookup should become a property function call (I think) is in `AttributeNode.analyse_attribute`. But it seems the entry is wrong, I need to either~ 

- ~fix the entry in `PropertyNode.analyse_declarations` and somehow make sure it is used instead of the old entry~
- ~change the code in `CFuncDefNode.analyse_declarations` and remove the `DecoratorTransform.visit_CFuncDefNode` function~

Any hints would be welcome.

Edit: added a new pass, ~still not working~

### CODE REVIEW

File: Cython/Compiler/Nodes.py [hunk 1]
```diff
@@ -2339,6 +2340,16 @@ def code_object(self):
...
+        if self.decorators:
+            for decorator in self.decorators:
+                func = decorator.decorator
+                if func.is_name:
+                    if func.name == 'classmethod' or func.name == 'staticmethod':
+                        error(self.pos, "Cannot handle these decorators yet")
+                    if func.name == 'property':
+                        # XXX DO SOMETHING HERE???
+                        pass
+
```
Comment (mattip) [inline Cython/Compiler/Nodes.py]: This XXX is hit if I remove `DecoratorTransform.visit_CFuncDefNode`, and the entry is registered properly for attribute lookup for `Foo.field0`, `Foo.field1`, `Foo.field2`, but then of course fails since the entry is a function `int (Foo)` not the function call.

File: Cython/Compiler/Nodes.py [hunk 2]
```diff
@@ -5023,6 +5034,13 @@ def analyse_declarations(self, env):
         self.entry = env.declare_property(self.name, self.doc, self.pos)
         self.entry.scope.directives = env.directives
         self.body.analyse_declarations(self.entry.scope)
+        # XXX DO SOMETHING HERE???
+        if 0 and self.is_wrapper:
+            entry = self.body.stats[0].entry
+            entry.is_property = 1
+            entry.doc = self.doc
+            env.property_entries[-1] = entry
+            env.entries[self.name] = entry
```
Comment (mattip) [inline Cython/Compiler/Nodes.py]: The entry from `env.declare_property` is not being overridden by the later entry from the body.stats[0].entry, even if I enable the code here

File: Cython/Compiler/Nodes.py [hunk 3]
```diff
@@ -2319,7 +2319,8 @@ class CFuncDefNode(FuncDefNode):
     #  is_static_method whether this is a static method
     #  is_c_class_method whether this is a cclass method
 
-    child_attrs = ["base_type", "declarator", "body", "py_func_stat"]
+    child_attrs = ["base_type", "declarator", "body", "py_func_stat", "decorators"]
+    outer_attrs = ["decorators"]
```
Comment (mattip) [inline Cython/Compiler/Nodes.py]: It is not clear to me how these are used, and if I need to add `'decorators'` to these.

File: Cython/Compiler/Nodes.py [hunk 4]
```diff
@@ -2339,6 +2340,16 @@ def code_object(self):
         return self.py_func.code_object if self.py_func else None
 
     def analyse_declarations(self, env):
+        if self.decorators:
+            for decorator in self.decorators:
+                func = decorator.decorator
+                if func.is_name:
+                    if func.name == 'classmethod' or func.name == 'staticmethod':
+                        error(self.pos, "Cannot handle these decorators yet")
+                    if func.name == 'property':
```
Comment (robertwb) [inline Cython/Compiler/Nodes.py]: We should probably error on everything but this for now.

File: Cython/Compiler/Nodes.py (hunk 1 above)
Comment (robertwb) [reply Cython/Compiler/Nodes.py]: We'd need a pass, likely somewhere before AnalyzeTypes, that rewrites all Foo.field0 into Foo.field0().

Comment (mattip) [reply Cython/Compiler/Nodes.py]: What is AnalyzeTypes? I turned on DebugFlags.debug_verbose_pipeline (since I am running a srctree test which runs cythonize in a subprocess, I needed to actually edit the file) and see a AnalyseDeclarationsTransform which is calling PropertyNode.analyse_declarations which is calling this function (CFuncDefNode.analyse_declarations) via line 5036 since it is part of the PropertyNode.body. But I couldn't figure out how to override the results of env.declare_property on line 5034 to make the CFuncDefNode's entry into the one called for `field0`

**Comment by mattip** [comment]
xref #2498 and numpy/numpy#11803

File: Cython/Compiler/ParseTreeTransforms.py [hunk 5]
```diff
@@ -1398,6 +1398,67 @@ def visit_DefNode(self, node):
         node.decorators = None
         return self.chain_decorators(node, decs, node.name)
 
+    def visit_CFuncDefNode(self, node):
```
Comment (scoder) [inline Cython/Compiler/ParseTreeTransforms.py]: I think it would be nice to reduce the code duplication here.

File: tests/run/ext_attr_getter.srctree [hunk 6]
```diff
@@ -78,32 +111,94 @@ cdef extern from "foo_nominal.h":
...
 
+######## getter1.pyx ########
+
+# Access base Foo fields from C via getter functions
+
+
+cdef extern from "foo.h":
+    ctypedef class foo_extension.Foo [object FooStructOpaque]:
+    # Importing will warn until we can use this syntax
+    # ctypedef class foo_extension.Foo [object FooStructOpaque, check_size]:
```
Comment (scoder) [inline tests/run/ext_attr_getter.srctree]: We can use this option now.

File: Cython/Compiler/Nodes.py (hunk 1 above)
Comment (mattip) [reply Cython/Compiler/Nodes.py]: Added a `ReplacePropertyNode` pass

File: Cython/Compiler/Nodes.py (hunk 3 above)
Comment (mattip) [reply Cython/Compiler/Nodes.py]: thanks

File: Cython/Compiler/ParseTreeTransforms.py (hunk 5 above)
Comment (mattip) [reply Cython/Compiler/ParseTreeTransforms.py]: removed it so the duplication is gone

File: tests/run/ext_attr_getter.srctree (hunk 6 above)
Comment (mattip) [reply tests/run/ext_attr_getter.srctree]: rebased off master and changed

**Comment by mattip** [comment]
I added a `ReplacePropertyNode` pass and verified that the `CFuncDefNode` was changed to a `SimpleCallNode`. However it seems the pass is too late or I need to propogate the changes further. 

The `AddNode` from `"f.field0 + f.field1"` has operands that are `AttributeNode`s, and those contain the old `CFuncDefNode`s. In which pass are the `AttributeNode`s updated or how can I trigger re-evaluating them?

**Comment by mattip** [comment]
It seems the "magic" is in getting from the `CDefExternNode.stats` to the  `CClassScope.entries`, there the `field0` is an `Entry` with an almost-correct but slightly incorrect type. So when the `AddNode` looks up its operands (in `AttributeNode.analyse_attribute`), the `AttributeNode` uses the `field0` entry, and fails.

File: Cython/Compiler/Nodes.py (hunk 4 above)
Comment (mattip) [reply Cython/Compiler/Nodes.py]: errors on anything except `property` and `staticmethod`

**Comment by mattip** [comment]
tests pass

File: Cython/Compiler/Nodes.py [hunk 7]
```diff
@@ -2411,11 +2421,16 @@ def analyse_declarations(self, env):
 
         type.is_const_method = self.is_const_method
         type.is_static_method = self.is_static_method
+        property = False
+        if self.decorators:
+            for _node in self.decorators:
+                if _node.decorator.is_name and _node.decorator.name == 'property':
```
Comment (scoder) [inline Cython/Compiler/Nodes.py]: The repetition of this looping code seems to scream for a helper function, e.g. one that searches for a specific decorator (and maybe applies some safety checks internally).

File: Cython/Compiler/ExprNodes.py [hunk 8]
```diff
@@ -6976,7 +6976,11 @@ def analyse_as_ordinary_attribute_node(self, env, target):
         self.obj = self.obj.analyse_types(env)
         self.analyse_attribute(env)
         if self.entry and self.entry.is_cmethod and not self.is_called:
-#            error(self.pos, "C method can only be called")
+            if getattr(self.entry, 'is_cgetter', False):
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py]: This should probably say `self.entry.type` – which suggests that this unused code is at least not tested.
Also, instead of using `getattr()` on a type, we normally set a class attribute somewhere in the type class hierarchy (or the `Entry`, or the `Node` hierarchy, …) and then override it with an instance attribute where necessary.

File: Cython/Compiler/ExprNodes.py [hunk 9]
```diff
@@ -11223,6 +11229,10 @@ def analyse_c_operation(self, env):
             self.operand2 = self.operand2.coerce_to(self.type, env)
 
     def compute_c_result_type(self, type1, type2):
+        if type1.is_cfunction and type1.is_cgetter:
+            type1 = type1.return_type
+        if type2.is_cfunction and type2.is_cgetter:
+            type2 = type2.return_type
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py]: I actually wonder if `is_cgetter` should be a feature of the type or of the entry. If the latter, then the code that calls this method should do the checking here. It doesn't really feel like a type specific thing to me, at least…

File: Cython/Compiler/Symtab.py [hunk 10]
```diff
@@ -829,14 +830,17 @@ def declare_cfunction(self, name, type, pos,
         type.entry = entry
         return entry
 
-    def add_cfunction(self, name, type, pos, cname, visibility, modifiers, inherited=False):
+    def add_cfunction(self, name, type, pos, cname, visibility, modifiers,
+                      inherited=False, property=False):
```
Comment (scoder) [inline Cython/Compiler/Symtab.py]: Maybe call it `is_property`, to avoid the builtin name?

File: tests/run/ext_attr_getter.srctree [hunk 11]
```diff
@@ -26,28 +33,55 @@ typedef struct {
...
+} FooStructOpaque;
+
+
+#define PyFoo_GET0M(a) ((FooStructNominal*)a)->f0
+#define PyFoo_GET1M(a) ((FooStructNominal*)a)->f1
+#define PyFoo_GET2M(a) ((FooStructNominal*)a)->f2
+
+int PyFoo_Get0F(FooStructOpaque *f)
+{
+    return ((FooStructNominal*)f)->f0;
```
Comment (scoder) [inline tests/run/ext_attr_getter.srctree]: Shouldn't this use the `#define` macros above?

File: tests/run/ext_attr_getter.srctree [hunk 12]
```diff
@@ -78,32 +111,89 @@ cdef extern from "foo_nominal.h":
...
+            return PyFoo_GET2M(self)
+
+    int PyFoo_GET0M(Foo);  # this is actually a macro !
+    int PyFoo_Get1F(Foo);
+    int PyFoo_GET2M(Foo);  # this is actually a macro !
+
+def sum(Foo f):
+    # Note - not a cdef function but compiling the f.__getattr__('field0')
+    # notices the getter and replaces the __getattr__ in c by PyFoo_GET anyway
+    return f.field0 + f.field1 + f.field2
```
Comment (scoder) [inline tests/run/ext_attr_getter.srctree]: We should test that the C level access is used here and not the Python level access. Not sure if [tree assertions](https://github.com/cython/cython/wiki/HackerGuide#parse-tree-assertions) work in srctree tests (they might not), but they seem necessary here, give that the Python interface is the same.

File: Cython/Compiler/ParseTreeTransforms.py [hunk 13]
```diff
@@ -2239,6 +2239,25 @@ def visit_IndexNode(self, node):
...
+                if len(node.decorators) > 1:
+                    return self._reject_decorated_property(node, decorator_node)
+                # Mark the node as a cgetter
+                node.type.is_cgetter = True
+                # Add a func_cname to be output instead of the attribute
+                node.entry.func_cname = node.body.stats[0].value.function.name
+                # done - remove the decorator node
+                node.decorators.remove(decorator_node)
+                return [node]
+        return [node]
```
Comment (scoder) [inline Cython/Compiler/ParseTreeTransforms.py]: I think returning the node (and not a single element list) should be fine here. Actually, I'd a) use `break` instead of duplicating the `return` code, and b) use a search loop that exits and then does stuff, rather than doing everything in the loop. (And c) use a helper to find the right decorator.)

File: Cython/Compiler/ExprNodes.py (hunk 8 above)
Comment (mattip) [reply Cython/Compiler/ExprNodes.py]: While the code path is hit, it is always False since `is_cgetter` lives on `self.entry.type`. It is cruft left over from a previous iteration of the code. Reverting.

File: Cython/Compiler/ExprNodes.py (hunk 9 above)
Comment (mattip) [reply Cython/Compiler/ExprNodes.py]: moved `c_getter` to `AttributeNode`

File: Cython/Compiler/Symtab.py (hunk 10 above)
Comment (mattip) [reply Cython/Compiler/Symtab.py]: ok

File: tests/run/ext_attr_getter.srctree (hunk 11 above)
Comment (mattip) [reply tests/run/ext_attr_getter.srctree]: either way ends up compiling to the same thing, adopting the idea macro use.

File: Cython/Compiler/ExprNodes.py (hunk 9 above)
Comment (mattip) [reply Cython/Compiler/ExprNodes.py]: It didn't work. The `@property` decorator is set on the `CFuncDefNode`. The `compute_c_result_type` is called for the `AttributeNode`. They both share a common `entry` instance. How can I transfer the `is_cgetter` set in `ReplacePropertyNode.visit_CFuncDefNode` to the `AttributeNode`that shares the entry?

File: Cython/Compiler/Nodes.py (hunk 7 above)
Comment (mattip) [reply Cython/Compiler/Nodes.py]: Merged with decorator analysis at the beginning of the function. Did you mean that I should refactor that into a helper function?

File: tests/run/ext_attr_getter.srctree (hunk 12 above)
Comment (mattip) [reply tests/run/ext_attr_getter.srctree]: I changed the attribute names in both the ctypedef class definition and `sum()`.

**Comment by mattip** [comment]
still outstanding review points that I need more guidance to complete: 
- [where `is_cgetter` should live](https://github.com/cython/cython/pull/2640#discussion_r245488503), on the `node` or on the `entry`, since two nodes reuse the same entry.
- [refactoring and a helper function](https://github.com/cython/cython/pull/2640#discussion_r245488799), refactored the code, but am not sure if this is the intent of the review comment
- [refactor](https://github.com/cython/cython/pull/2640#discussion_r245489153) the node property analysis, refactored but I am not sure if this is the intent of the review comment

File: Cython/Compiler/ExprNodes.py (hunk 9 above)
Comment (scoder) [reply Cython/Compiler/ExprNodes.py]: Through the `Entry`. :) Symbol table entries are shared within their scope.
Add a class attribute `is_cgetter = False` to the `Entry` class and then set it to `True` on the instance where you need it.

File: Cython/Compiler/ExprNodes.py [hunk 14]
```diff
@@ -6976,11 +6976,7 @@ def analyse_as_ordinary_attribute_node(self, env, target):
         self.obj = self.obj.analyse_types(env)
         self.analyse_attribute(env)
         if self.entry and self.entry.is_cmethod and not self.is_called:
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py]: Maybe we should only set `entry.is_cgetter` and not `entry.is_cmethod` for these properties ? If `@propertiy` is used, I think it's intended to make the entry non-callable, i.e. not a method.

File: Cython/Compiler/Nodes.py [hunk 15]
```diff
@@ -2340,11 +2340,14 @@ def code_object(self):
...
 
     def analyse_declarations(self, env):
+        is_property = False
         if self.decorators:
             for decorator in self.decorators:
                 func = decorator.decorator
                 if func.is_name:
-                    if func.name in ('property', 'staticmethod'):
+                    if func.name == 'property':
+                        is_property = True
```
Comment (scoder) [inline Cython/Compiler/Nodes.py]: Ah, yes, that's much better.

File: Cython/Compiler/ParseTreeTransforms.py [hunk 16]
```diff
@@ -2245,18 +2245,23 @@ def visit_CFuncDefNode(self, node):
...
-            if decorator.is_name and decorator.name == 'property':
-                if len(node.decorators) > 1:
-                    return self._reject_decorated_property(node, decorator_node)
-                # Mark the node as a cgetter
-                node.type.is_cgetter = True
-                # Add a func_cname to be output instead of the attribute
-                node.entry.func_cname = node.body.stats[0].value.function.name
-                # done - remove the decorator node
+            _node = self.analyse_decorator(node, decorator_node.decorator)
+            if _node:
```
Comment (scoder) [inline Cython/Compiler/ParseTreeTransforms.py]: Hmm, this refactoring doesn't look helpful. The return value means different things inside of the `analyse_decorator()` method, doesn't have an explanatory name here, and is basically ignored. That's rather obscure code.
Also, `analyse_decorator()` essentially consists of two code paths, but I have to look through the longest path in order to understand the shorter one (the one that says "nothing to do").

So, first of all, I think this whole transformation can be merged into the `DecoratorTransform`. Then, the `visit_DefNode()` method over there should share some code with `visit_CFuncDefNode()`. They will be doing different things in the end, but the search loop, conditions and error handling are the same, so they should only be implemented once. I propose to write a shared helper method that searches an `@property` (or `@setter` etc.) decorator, and, if found (and it's the only one), returns it, so that the two visitor methods can deal with it in their own ways. For all errors, it should just call `error()`. Failing to find a property decorator, returning `None` is a good enough signal to the caller.

Note that `_reject_decorated_property()` actually raises a `CompileError` in the `error()` call, so it does not actually return (I admit that that's not obvious either…).

File: Cython/Compiler/ExprNodes.py (hunk 9 above)
Comment (mattip) [reply Cython/Compiler/ExprNodes.py]: Confusing. In order to move the `is_getter` off the entry, I have to set it on the entry anyway? Wouldn't it be simpler to just leave it on the entry?

File: Cython/Compiler/ParseTreeTransforms.py (hunk 16 above)
Comment (mattip) [reply Cython/Compiler/ParseTreeTransforms.py]: We reached the conclusion that an extra pass was needed in one of the previous iterations of this PR, since I need to mark the `node.type` attribute (which is an `Entry`), and that is calculated in the `AnalyseDeclarationsTransform` pass, which happens after the `DecoratorTransform` pass.

File: Cython/Compiler/ExprNodes.py (hunk 14 above)
Comment (mattip) [reply Cython/Compiler/ExprNodes.py]: it is still callable, just as a c-function not a method. I tried setting `is_cmethod =0` when `is_cgetter = True` and it broke the tests, requiring more logic everywhere `is_cmethod` is checked.

File: Cython/Compiler/ParseTreeTransforms.py (hunk 16 above)
Comment (scoder) [reply Cython/Compiler/ParseTreeTransforms.py]: Sorry, my bad. The usual problem of long-living PRs/branches/designs/…

File: Cython/Compiler/ParseTreeTransforms.py [hunk 17]
```diff
@@ -2243,23 +2243,24 @@ class ReplacePropertyNode(CythonTransform):
     def visit_CFuncDefNode(self, node):
         if not node.decorators:
             return node
-        # transform @property decorators on ctypedef class functions
+        decorator = self.find_first_decorator(node, 'property')
```
Comment (scoder) [inline Cython/Compiler/ParseTreeTransforms.py]: Yep, I think that's better.

File: Cython/Compiler/Nodes.py (hunk 2 above)
Comment (scoder) [reply Cython/Compiler/Nodes.py]: What is the intention here?

File: Cython/Compiler/Symtab.py [hunk 18]
```diff
@@ -752,7 +754,8 @@ def register_pyfunction(self, entry):
 
     def declare_cfunction(self, name, type, pos,
                           cname=None, visibility='private', api=0, in_pxd=0,
-                          defining=0, modifiers=(), utility_code=None, overridable=False):
+                          defining=0, modifiers=(), utility_code=None,
+                          overridable=False, is_property=False):
```
Comment (scoder) [inline Cython/Compiler/Symtab.py]: Since the flag is unused here, maybe we should add an assertion that it's not set to true anywhere?
Or, actually, not have it here at all?

File: Cython/Compiler/Nodes.py (hunk 2 above)
Comment (mattip) [reply Cython/Compiler/Nodes.py]: sorry, removing old cruft.

File: Cython/Compiler/Symtab.py [hunk 19]
```diff
@@ -1933,7 +1940,8 @@ def declare_var(self, name, type, pos,
 
     def declare_cfunction(self, name, type, pos,
                           cname=None, visibility='private', api=0, in_pxd=0,
-                          defining=0, modifiers=(), overridable=False):  # currently no utility code ...
+                          defining=0, modifiers=(), overridable=False,
+                          is_property=False):  # currently no utility code ...
```
Comment (scoder) [inline Cython/Compiler/Symtab.py]: Same thing here, I'd rather handle `is_property` the same as the `utility_code` argument for scopes that do not support it, i.e. make it a hard error if we end up passing it, by simply not allowing it.

Comment (mattip) [reply Cython/Compiler/Symtab.py]: The call chain is `Nodes.CFuncDefNode.analyse_declarations` finds the property decorator and later calls `env.declare_cfunction` where `env` is any one of `Scope`, `CClassScope` or .`CppClassScope`. If I remove the kwarg from `Symtab.Scope`, how do I adjust the call in `analyse_declarations`?

Also I see that I never handle the kwarg in `Symtab.CppClassScope.declare_cfunction`. Do I need to add a test for a `CppClass` that uses a `@property to define a getter? I didn't really grok that `CppClass` is different from `CClass`

File: Cython/Compiler/Symtab.py (hunk 18 above)
Comment (mattip) [reply Cython/Compiler/Symtab.py]: In the test, `Scope.declare_cfunction` is called once from `CFuncDefNode.analyse_declarations` for `__pyx_unpickle_Foo__set_state`. Removing the kwarg means checking the `env` type before the call

File: Cython/Compiler/Symtab.py (hunk 19 above)
Comment (mattip) [reply Cython/Compiler/Symtab.py]: After staring at the examples of `cdef cppclass` I think use of a `@property` decorator to replace a `cdef attribute` getter with a c function is not needed, so the code path in `Symtab.CppClassScope.declare_cfunction` should never have `is_property` be True. Do you agree, and should I assert `is_property == False` at the top of this function?

Comment (mattip) [reply Cython/Compiler/Symtab.py]: ok, removed the kwarg, at the cost of moving the entry from one property list to another after its creation

Comment (mattip) [reply Cython/Compiler/Symtab.py]: just now, 261ad4d

File: tests/run/cpp_class_attrib.srctree [hunk 20]
```diff
@@ -0,0 +1,78 @@
+PYTHON setup.py build_ext --inplace
+PYTHON -c "import runner"
+
+######## setup.py ########
+
+from Cython.Build.Dependencies import cythonize
+from distutils.core import setup
+import os
+
+examples = '../../../docs/examples/userguide/wrapping_CPlusPlus'
```
Comment (scoder) [inline tests/run/cpp_class_attrib.srctree]: I would prefer having a simple version of the Rectangle C++ source here, rather than looking it up in some far away directory through brittle relative paths. You can use [verbatim C code](http://docs.cython.org/en/latest/src/userguide/external_C_code.html#including-verbatim-c-code) for that.

Also, I'm currently enabling the C++ tests on appveyor (#2797). Once that works and I merged it, you can merge the latest master into your feature branch to let appveyor run this test.

Comment (mattip) [reply tests/run/cpp_class_attrib.srctree]: Whoops, I didn't mean for the cpp attribute example to creep in. The feature I am adding is only for `ctypedef`-defined classes.

File: Cython/Compiler/Nodes.py [hunk 21]
```diff
@@ -2411,11 +2424,16 @@ def analyse_declarations(self, env):
...
+
         self.entry = env.declare_cfunction(
             name, type, self.pos,
             cname=cname, visibility=self.visibility, api=self.api,
             defining=self.body is not None, modifiers=self.modifiers,
             overridable=self.overridable)
+        if is_property:
+            self.entry.is_property = 1
+            env.property_entries.append(self.entry)
+            env.cfunc_entries.remove(self.entry)
```
Comment (scoder) [inline Cython/Compiler/Nodes.py]: This makes me wonder why we shouldn't be calling `declare_property()` here in the first place, instead of `declare_cfunction()`. What do you think? It seems to do the right thing, pretty much.

File: Cython/Compiler/Symtab.py [hunk 22]
```diff
@@ -752,7 +754,8 @@ def register_pyfunction(self, entry):
 
     def declare_cfunction(self, name, type, pos,
                           cname=None, visibility='private', api=0, in_pxd=0,
-                          defining=0, modifiers=(), utility_code=None, overridable=False):
+                          defining=0, modifiers=(), utility_code=None,
+                          overridable=False):
```
Comment (scoder) [inline Cython/Compiler/Symtab.py]: These are now unfortunate, unnecessary changes. Could you undo the signature reformatting?

File: Cython/Compiler/Nodes.py (hunk 21 above)
Comment (mattip) [reply Cython/Compiler/Nodes.py]: `CClassScope.declare_property` does not dive into the wrapped function the way `CClassScope.declare_cfuntion` does. I played with this for a while but could not get it to work cleanly, without in the end calling `declare_cfunction` anyway.

File: Cython/Compiler/Symtab.py (hunk 22 above)
Comment (mattip) [reply Cython/Compiler/Symtab.py]: done

File: Cython/Compiler/Nodes.py (hunk 21 above)
Comment (mattip) [reply Cython/Compiler/Nodes.py]: done, extra test added. Since I want to print the `type` of the unhandled instance, I needed to rename `type` -> `type` in that function.

File: Cython/Compiler/Nodes.py [hunk 23]
```diff
@@ -2339,6 +2340,18 @@ def code_object(self):
...
+        if self.decorators:
+            for decorator in self.decorators:
+                func = decorator.decorator
+                if func.is_name:
+                    if func.name == 'property':
+                        is_property = 1
+                    elif func.name == 'staticmethod':
+                        pass
+                    else:
+                        error(self.pos, "Cannot handle %s decorators yet" % func.name)
```
[the remark this answers is not in the document]
Comment (mattip) [reply Cython/Compiler/Nodes.py]: done, extra test added. Since I want to print the `type` of the unhandled instance, I needed to rename `type` -> `type` in that function.

File: Cython/Compiler/ExprNodes.py [hunk 24]
```diff
@@ -7146,6 +7146,8 @@ def calculate_result_code(self):
         obj_code = obj.result_as(obj.type)
         #print "...obj_code =", obj_code ###
         if self.entry and self.entry.is_cmethod:
+            if self.entry.is_cgetter:
+                return "%s(%s)" %(self.entry.func_cname, obj_code)
```
[the remark this answers is not in the document]
Comment (mattip) [reply Cython/Compiler/ExprNodes.py]: done

**Comment by mattip** [comment]
How do you easily tell which test is failing on travis? I ended up rerunning the failing build locally

File: Cython/Compiler/Symtab.py (hunk 18 above)
Comment (mattip) [reply Cython/Compiler/Symtab.py]: reworked to avoid adding the kwarg

**Comment by scoder** [comment]
Looks ok now, thanks!

**Comment by scoder** [comment]
> a decorated `CFuncDefNode` is different from a `Property` node

Yeah, that sounds wrong. We should reuse the property infrastructure that we have. IIUC, the only difference is that these properties are resolved at compile time and the underlying method gets called. They are essentially "inline" properties.

**Comment by mattip** [comment]
Is there any assumption Property nodes are on python classes and not c classes? Or maybe that's what you mean about "inline".

**Comment by scoder** [comment]
They are always on `cdef` classes. Properties on Python classes are just arbitrary objects and not special syntax tree constructs.
The missing feature is that property nodes were not previously useful on external classes, since there was no code that Cython needed to generate for them. The fact that the usage of properties could be inlined into direct method calls (not only for external classes) was not really considered. But it could apply to properties marked as `@cython.inline` as well as any property on final classes, whether implemented in the module or externally cimported.

**Comment by scoder** [comment]
(BTW, thanks for asking the right questions and digging into the details of this feature. I think we are slowly getting an idea of where this is really going.)

**Comment by mattip** [comment]
I want to get back to this and finish it up.

> But it could apply to properties marked as @cython.inline as well as any property on final classes, whether implemented in the module or externally cimported

I am not sure what you mean by this. Looking at the implementation of `@cython.inline`, it seems more like a construct used in a `.py` file to create a "local" c-extension module with a single function than a general-purpose decorator on a `cdef` class or function
